### PR TITLE
feat: add gRPC-based NodeExecutorService

### PR DESCRIPTION
**Pull request by JKRhb**
This PR will eventually add the missing code for communicating with the `NodeExecutor` from a `NodeInstance` via gRPC.

### CODE REVIEW

**Comment by JKRhb** [comment]
@pulsastrix I think this PR should now be ready for an initial review. In the most recent commits, I tried to achieve a loose coupling by defining two interfaces (`ExternalExecutor` and `CongestionReporter`).

The former of the two interfaces can be implemented by external applications/libraries to be triggered by the Iceflow system, while the latter of the two is used to trigger the congestion reporter from within the `IceflowProducer` class. I am not yet sure where to put the logic to trigger the congestion reporting, but we can either discuss that in the context of this PR or in a follow-up one.

Note that the commit history is not that clean yet, I will reorganize the commits once the PR has been approved.

File: include/iceflow/executor.hpp
Comment (pulsastrix) [inline include/iceflow/executor.hpp:44]: Ideally, the congestion report would also include information regarding the edge that is congested, although it is not strictly necessary.

_Usually_, we would detect that an edge in the graph is congested based on the difference between produced and consumed units summed over all nodes.
However, this only works based on the assumption that all nodes are equal and that there is no network congestion.
In any other case, only some of the instances of a given task might be overloaded, and in those cases, it might be good to explicitly provide information about the congested edge to the node executor.

**Comment by pulsastrix** [comment]
I just noticed that we're still missing an RPC call from the executor to the instance to query for production/consumption stats for each node. We should address that (probably in a followup PR) as it is required for the scaling algorithm.

**Comment by JKRhb** [comment]
> I just noticed that we're still missing an RPC call from the executor to the instance to query for production/consumption stats for each node. We should address that (probably in a followup PR) as it is required for the scaling algorithm.

I've just created #82 as a reminder, feel free to adjust it as needed.

File: include/iceflow/executor.hpp
Comment (pulsastrix) [reply include/iceflow/executor.hpp]: Looks good for now, although I am unsure if we really want to define a name (in string form) for each edge separately.

Comment (JKRhb) [reply include/iceflow/executor.hpp]: Hmm, do you think a numeric identifier would make more sense here then?

Comment (JKRhb) [reply include/iceflow/executor.hpp]: Or do you mean that we should make the field/parameter optional?

**Comment by JKRhb** [comment]
The PR should now be ready for a final review :) If the changes are fine for you, feel free to just hit the merge button :)

File: proto/iceflow.proto
Comment (pulsastrix) [inline proto/iceflow.proto:63]: I don't think so, at least for now.

File: proto/iceflow.proto
Comment (pulsastrix) [inline proto/iceflow.proto:67]: Those two should be sufficient for now, considering that it is the producer that sends these congestion notifications, and the producer doesn't really have more information than "my queue is filling up" (processing) or "i received an NDN congestion notification" (network).

File: src/producer.cpp
Comment (pulsastrix) [inline src/producer.cpp:112]: We'd probably have to integrate with NDN-SVS in order to manage the queue for a given edge (both to delete already processed elements and to detect processing congestions based on the queue size).

**Review by pulsastrix — APPROVED** [review]
LGTM, the new comments are just my two cents on the remaining TODOs (which are probably better addressed in new PRs).

File: include/iceflow/executor.hpp
Comment (pulsastrix) [reply include/iceflow/executor.hpp]: Ideally, we'd have some identifier that can be automatically generated based on the start and end node of the DAG (so that we don't need to generate more identifiers), but that might cause problems for some graphs (e.g. if there are multiple edges between the same two nodes).

I'd keep it as a string for now, maybe we can change that later though.

File: proto/iceflow.proto
Comment (JKRhb) [reply proto/iceflow.proto]: Thank you for the feedback on this :) I removed the TODO and added some additional documentation to individual reasons.

File: proto/iceflow.proto
Comment (JKRhb) [reply proto/iceflow.proto]: Thanks, I removed the TODO via b1f352e8292ba2c2a1b4e1f2b956c0408aa87379 :)

File: src/producer.cpp
Comment (JKRhb) [reply src/producer.cpp:112]: Yeah, that makes sense. Should we create an issue to keep track of that or do you think keeping the TODO wold suffice here?

Comment (pulsastrix) [reply src/producer.cpp:112]: Keeping the TODO will probably suffice.

**Pull request merged**
